### ISSUE TITLE
Problem with custom AMI

### ISSUE THREAD

**Issue by acastanedam** [aws/aws-parallelcluster#2199]
**Required Info:**
 - AWS ParallelCluster version 2.9.1:
 - Full cluster configuration without any credentials or personal data
`
[global]
cluster_template = gpu
sanity_check = true
update_check = true

[aws]
aws_region_name = eu-central-1

[aliases]
ssh = ssh {CFN_USER}@{MASTER_IP} {ARGS}

[cluster gpu]
base_os = ubuntu1804
compute_root_volume_size = 250
custom_ami = ami-03b89d989e60a4cda
dcv_settings = custom
ebs_settings = custom
enable_intel_hpc_platform = false
extra_json = { "cluster" : { "skip_install_recipes" : "no" } }
key_name = *
master_instance_type = p3.8xlarge
master_root_volume_size = 250
queue_settings = custom
scaling_settings = custom
scheduler = slurm
tags = {"Name": "TNT", "OS": "ubuntu"}
vpc_settings = custom

[compute_resource custom]
initial_count = 1
instance_type = p3.8xlarge
max_count = 2
min_count = 1

[dcv custom]
enable = master

[ebs custom]
shared_dir = /shared
volume_size = 2400

[queue custom]
compute_resource_settings = custom
compute_type = ondemand
disable_hyperthreading = true
enable_efa = false
placement_group = DYNAMIC

[scaling custom]
scaledown_idletime = 10

[vpc custom]
vpc_id = vpc-36a75a5f
master_subnet_id = subnet-b69c7dcd`

 - Cluster name: gpitnt

**Bug description and how to reproduce:**
Trying to create a cluster with p3.8xlarge instances and an AWS AMI for machine learning with:
`pcluster create -c .aws/pcl_config/frankfurt_pcluster_config_gpu -t gpu -nr gpitnt`
ends in
`Beginning cluster creation for cluster: gpitnt
Creating stack named: parallelcluster-gpitnt
Status: parallelcluster-gpitnt - CREATE_FAILED                                  
Cluster creation failed.  Failed events:
  - AWS::CloudFormation::Stack parallelcluster-gpitnt The following resource(s) failed to create: [MasterServerSubstack]. 
  - AWS::CloudFormation::Stack MasterServerSubstack Embedded stack arn:aws:cloudformation:eu-central-1:851368960188:stack/parallelcluster-gpitnt-MasterServerSubstack-1RZHYG5KEYJUC/856e5eb0-1e97-11eb-a68c-02205d104246 was not successfully created: The following resource(s) failed to create: [MasterServer]. 
    - AWS::CloudFormation::Stack parallelcluster-gpitnt-MasterServerSubstack-1RZHYG5KEYJUC The following resource(s) failed to create: [MasterServer]. 
    - AWS::EC2::Instance MasterServer Failed to receive 1 resource signal(s) within the specified duration`

The cluster is working however (with `-nr` of course). According to the message looks like a time threshold issue.

**If you are reporting issues about cluster creation failure or node failure:**

These are the logfiles
[pcluster_logs.tar.gz](https://github.com/aws/aws-parallelcluster/files/5487967/pcluster_logs.tar.gz)

Thanks in advance for your help,

Arcesio Castaneda M.

**Comment by demartinofra** [comment]
Hi Arcesio,

the timeout is being triggered because of the slow provisioning time of the head node and the compute node necessary to satisfy min queue capacity. The reason behind the slow provisioning is that you are using a plain Deep Learning AMI which is missing all ParallelCluster components. Because of this ParallelCluster ends up performing the installation of all components at runtime, resulting in a very slow provisioning.
What I suggest you to do is to follow [this guide](https://docs.aws.amazon.com/parallelcluster/latest/ug/tutorials_02_ami_customization.html#build-a-custom-aws-parallelcluster-ami) in order to build a custom AMI that starts from the Deep Learning one and pre-bakes all ParallelCluster bits in it. Once you have this AMI you can simply use in place of the DL AMI you are using today. This has the advantage to reduce the node launch time from ~20 mins to ~3 mins, resulting in a more efficient utilization of resources and cost savings.
If you still want to stick to the runtime installation approach you could set the min_count and initial_count to 0 so that the Cluster is able to complete creation before the timeout is triggered. Unfortunately today increasing this timeout requires a change in the ParallelCluster templates and it's not a recommended path to follow.

Please let us know if you need any additional guidance in creating your custom AMI starting from the DL AMI.

Cheers,
Francesco

**Comment by acastanedam** [comment]
Hi Francesco,

Thanks for the information. The problem is when I tried to create the custom AMI there are also problems.
With:
`pcluster createami -ai ami-091edcc9f1543188f -os ubuntu1804 -i p3.8xlarge -c .aws/pcl_config/frankfurt_pcluster_config_gpu --region eu-central-1`
the process ends without success:
`
Building AWS ParallelCluster AMI. This could take a while...                                                                                                                                  
Base AMI ID: ami-091edcc9f1543188f                                                                                                                                                            
Base AMI OS: ubuntu1804
Instance Type: p3.8xlarge                                                                      
Region: eu-central-1                                                                           
VPC ID: vpc-36a75a5f                                                                           
Subnet ID: subnet-b69c7dcd                                                                     
Template: https://eu-central-1-aws-parallelcluster.s3.eu-central-1.amazonaws.com/templates/aws-parallelcluster-2.9.1.cfn.json
Cookbook: https://eu-central-1-aws-parallelcluster.s3.eu-central-1.amazonaws.com/cookbooks/aws-parallelcluster-cookbook-2.9.1.tgz
Packer log: /tmp/packer.log.20201104-190509.zauwzzlu                          
Packer Instance ID: i-0b9c5aadbf32f6de8                                                        
Packer status:  exit code 1       
No custom AMI created
`
Note, here `ami-091edcc9f1543188f ` is already a custom AMI based on the AWS AMI `ami-03b89d989e60a4cda` which I tried to use in the above config. file. This means I only changed the `custom_ami` parameter.

One further question, is it possible to change (at the user level) the time limit you mentioned above.

I attach the corresponding logfile.
[packer.log](https://github.com/aws/aws-parallelcluster/files/5490229/packer.log)

Best,
Arcesio

**Comment by demartinofra** [comment]
By quickly looking at the packer log I can see that it failed when trying to ssh to the instance needed to create the custom AMI. Can you check if you have any rule configured at the VPC/Subnet level (VPC ID: vpc-36a75a5f Subnet ID: subnet-b69c7dcd) that is restricting access and also can you confirm you can successfully launch and ssh into an instance by using the ami ami-091edcc9f1543188f? From what I can see this is not the vanilla DLAMI anymore right?

***
```
==> custom-ubuntu1804: Prevalidating any provided VPC information
==> custom-ubuntu1804: Prevalidating AMI Name: custom-ami-aws-parallelcluster-2.9.1-ubuntu-1804-lts-hvm-x86_64-202011041905
custom-ubuntu1804: Found Image ID: ami-091edcc9f1543188f
==> custom-ubuntu1804: Creating temporary keypair: packer_5fa2ed56-b44e-d07b-bd40-233c202b89a6
==> custom-ubuntu1804: Creating temporary security group for this instance: packer_5fa2ed57-51cd-1a22-9b13-7931f2dcffb8
==> custom-ubuntu1804: Authorizing access to port 22 from [0.0.0.0/0] in the temporary security groups...
==> custom-ubuntu1804: Launching a source AWS instance...
==> custom-ubuntu1804: Adding tags to source instance
custom-ubuntu1804: Adding tag: "build_date": "202011041905"
custom-ubuntu1804: Adding tag: "Name": "Packer Builder ami-091edcc9f1543188f"
custom-ubuntu1804: Adding tag: "parallelcluster_version": "aws-parallelcluster-2.9.1"
custom-ubuntu1804: Instance ID: i-0b9c5aadbf32f6de8
==> custom-ubuntu1804: Waiting for instance (i-0b9c5aadbf32f6de8) to become ready...
==> custom-ubuntu1804: Using ssh communicator to connect: 18.193.254.117
==> custom-ubuntu1804: Waiting for SSH to become available...
==> custom-ubuntu1804: Timeout waiting for SSH.
==> custom-ubuntu1804: Terminating the source AWS instance...
==> custom-ubuntu1804: Cleaning up any extra volumes...
==> custom-ubuntu1804: No volumes to clean up, skipping
==> custom-ubuntu1804: Deleting temporary security group...
==> custom-ubuntu1804: Deleting temporary keypair...
Build 'custom-ubuntu1804' errored after 11 minutes 50 seconds: Timeout waiting for SSH.
```

**Comment by acastanedam** [comment]
Hi Franceso,
Right this is not anymore the "pure" DLAMI, although I have already tried the originak AWS DLAMIN and the same happens. Using an AWS AMI not in the list of aws-parallelcluster is not recommended, right?
On the other hand, I can log into such instance, actually I can "create" a cluster and log into the master if I use `-nr`, in such case the same problem described in the first message also occurs.
Thanks,
Arcesio

**Comment by tilne** [comment]
@acastanedam To clarify: I think in his last comment Francesco was suggesting a possible cause for the `createami` failure as opposed to the cluster creation failure.

Were you able to try launching an EC2 instance using AMI ami-091edcc9f1543188f in subnet subnet-b69c7dcd of VPC vpc-36a75a5f? I believe the purpose of trying this out is to make sure that there's nothing at the subnet/VPC level or any software changes in your custom AMI that would prevent packer from SSH'ing to an instance it creates in that subnet when creating the AMI.

> Using an AWS AMI not in the list of aws-parallelcluster is not recommended, right?

I'm not sure what you're asking. I think you're asking if using an AMI other than the defaults created by ParallelCluster is not recommended? If so, I would say that we definitely support using non-default AMIs, but we recommend installing the required ParallelCluster software on top of those non-default AMIs using `pcluster createami`. I know you and Francesco discussed this above, so please let me know if I'm answering the wrong question.

**Comment by acastanedam** [comment]
Hi,
I have just tried to create the sinlge instance with AMI ami-091edcc9f1543188f in subnet subnet-b69c7dcd of VPC vpc-36a75a5f and it worked, and can ssh into it.
About the question, I think I put it in the wrong way, sorry. I was referring to:
`You shouldn't use an AWS ParallelCluster AMI as <BASE_AMI> for the createami command. Otherwise, the command fails.` in the instructions  (`Build a Custom AWS ParallelCluster AMI`) for `createami`.
On the other hand, I the problem was due to the fact I was using `pcluster createami` from a machine which actually requires to go trough another one to be able to do `ssh` to the AWS services. As it is not possible to append commands to `ssh` for `createami` (as in `pcluster ssh`) I have to move to the authorized machine and create the AMI there. This works and now I can use the output AMI to create and use the cluster from any of those machines.
I do not know if it will be useful to consider adding this `pcluster ssh` feature also to `pcluster createami`

Thanks for your comments and help
